libr/anal/p/anal_evm.c

```diff
@@ -0,0 +1,331 @@
+	[EVM_OP_SELFDESTRUCT] = R_ANAL_OP_TYPE_CRYPTO, 
+};
+
+static int evm_oplen(ut8 opcode) {
+    int ret;
+	EvmOpDef *opdef = &opcodes[opcode];
+
+	if (opdef->txt) {
+		return opdef->len;
+	}
+	switch (opcode) {
+	case EVM_OP_PUSH1:
+	case EVM_OP_PUSH2:
+	case EVM_OP_PUSH3:
+	case EVM_OP_PUSH4:
+	case EVM_OP_PUSH5:
+	case EVM_OP_PUSH6:
+	case EVM_OP_PUSH7:
+	case EVM_OP_PUSH8:
+	case EVM_OP_PUSH9:
+	case EVM_OP_PUSH10:
+	case EVM_OP_PUSH11:
+	case EVM_OP_PUSH12:
+	case EVM_OP_PUSH13:
+	case EVM_OP_PUSH14:
+	case EVM_OP_PUSH15:
+	case EVM_OP_PUSH16:
+	case EVM_OP_PUSH17:
+	case EVM_OP_PUSH18:
+	case EVM_OP_PUSH19:
+	case EVM_OP_PUSH20:
+	case EVM_OP_PUSH21:
+	case EVM_OP_PUSH22:
+	case EVM_OP_PUSH23:
+	case EVM_OP_PUSH24:
+	case EVM_OP_PUSH25:
+	case EVM_OP_PUSH26:
+	case EVM_OP_PUSH27:
+	case EVM_OP_PUSH28:
+	case EVM_OP_PUSH29:
+	case EVM_OP_PUSH30:
+	case EVM_OP_PUSH31:
+	case EVM_OP_PUSH32:
+		{
+			int i, pushSize = opcode - EVM_OP_PUSH1;
+            /*
+			op->imm = 0;
+			for (i = 0; i < pushSize + 1; i++) {
+				op->imm <<= 8;
+				op->imm |= buf[i + 1];
+			}
+			settxtf (op, "push%d 0x%x", pushSize + 1, op->imm);
+            */
+			ret = 2 + pushSize;
+		}
+		break;
+	case EVM_OP_DUP1:
+	case EVM_OP_DUP2:
+	case EVM_OP_DUP3:
+	case EVM_OP_DUP4:
+	case EVM_OP_DUP5:
+	case EVM_OP_DUP6:
+	case EVM_OP_DUP7:
+	case EVM_OP_DUP8:
+	case EVM_OP_DUP9:
+	case EVM_OP_DUP10:
+	case EVM_OP_DUP11:
+	case EVM_OP_DUP12:
+	case EVM_OP_DUP13:
+	case EVM_OP_DUP14:
+	case EVM_OP_DUP15:
+	case EVM_OP_DUP16:
+		{
+			int dupSize = opcode - EVM_OP_DUP1 + 1;
+			//settxtf (op, "dup%d", dupSize);
+			ret = dupSize + 1;
+		}
+		break;
+	case EVM_OP_SWAP1:
+	case EVM_OP_SWAP2:
+	case EVM_OP_SWAP3:
+	case EVM_OP_SWAP4:
+	case EVM_OP_SWAP5:
+	case EVM_OP_SWAP6:
+	case EVM_OP_SWAP7:
+	case EVM_OP_SWAP8:
+	case EVM_OP_SWAP9:
+	case EVM_OP_SWAP10:
+	case EVM_OP_SWAP11:
+	case EVM_OP_SWAP12:
+	case EVM_OP_SWAP13:
+	case EVM_OP_SWAP14:
+	case EVM_OP_SWAP15:
+	case EVM_OP_SWAP16:
+		{
+			int swapSize = opcode - EVM_OP_SWAP1 + 1;
+			//settxtf (op, "swap%d", swapSize);
+			ret = 1;
+		}
+		break;
+	case EVM_OP_LOG0:
+	case EVM_OP_LOG1:
+	case EVM_OP_LOG2:
+	case EVM_OP_LOG3:
+	case EVM_OP_LOG4:
+		{
+			int logSize = opcode - EVM_OP_LOG0;
+			//settxtf (op, "log%d", logSize);
+			ret = 1;
+		}
+		break;
+	default:
+		//settxtf (op, "invalid");
+		ret = 1;
+		break;
+	}
+
+	return ret;
+}
+
+static int evm_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len) {
+    static ut16 a = 0;
+    int ret;
+    ut8 opcode;
+
+    opcode = buf[0];
+    anal->addrmy  = op->addr;
+
+    memset(op, 0, sizeof(RAnalOp));
+
+    op->type = opcodes_types[opcode];
+	if (!op->type) {
+		op->type = R_ANAL_OP_TYPE_UNK;
+	}
+    op->addr = addr;
+    op->jump = op->fail = -1;
+    op->ptr = op->val = -1;
+
+    r_strbuf_init (&op->esil);
+
+    switch(opcode) {
+    case EVM_OP_JUMP:
+        op->fail = addr+1;
+        op->jump = buf[-1];
```

Comment:
You need to be sure that `buf[0] - 1` is a valid address, what if it is a beginning of a buffer?